### PR TITLE
push triggers pull, fix glitches

### PR DESCRIPTION
**Pull request by limemloh**
This should PR changes how push works with Behaviors. 
Push now triggers a pull, which ensures that all the dependencies are calculated first, eliminating glitches in Behaviors.

I have added two simple tests showing exposing glitches.

### CODE REVIEW

File: src/behavior.ts [hunk 1]
```diff
@@ -69,27 +69,25 @@ export abstract class Behavior<A> extends Reactive<A, BListener>
...
     if (this.pulledAt === undefined || this.pulledAt < t) {
       this.pulledAt = t;
       let shouldRefresh = this.changedAt === undefined;
       for (const parent of this.parents) {
-        if (isBehavior(parent)) {
-          if (parent.state !== State.Push && parent.pulledAt !== t) {
-            parent.pull(t);
-          }
-          shouldRefresh = shouldRefresh || parent.changedAt > this.changedAt;
+        if (!isBehavior(parent)) {
```
Comment (paldepind) [inline src/behavior.ts]: Why negate the check here? The code inside the `for`-loop could be:

```ts
if (isBehavior(parent)) {
  parent.pull(t);
  shouldRefresh = shouldRefresh || this.changedAt < parent.changedAt;
}
```

File: src/behavior.ts [hunk 2]
```diff
@@ -128,39 +126,44 @@ export function pushToChildren(t: number, b: Behavior<any>): void {
   }
 }
 
-function refresh<A>(b: Behavior<A>, t: number): void {
+function refresh<A>(b: Behavior<A>, t: number) {
   const newValue = b.update(t);
-  if (newValue !== b.last) {
-    b.changedAt = t;
-    b.last = newValue;
+  if (newValue === b.last) {
```
Comment (paldepind) [inline src/behavior.ts]: Why negate the check here instead of the old code?

File: src/placeholder.ts [hunk 3]
```diff
@@ -57,6 +54,7 @@ export class Placeholder<A> extends Behavior<A> {
     if (this.source !== undefined) {
       this.source.addListener(this.node, t);
       if (isBehavior(this.source)) {
+        this.pull(t);
```
Comment (paldepind) [inline src/placeholder.ts]: Why is this pull there? As far as I can see all the tests still pass if I comment it out.

File: src/behavior.ts [hunk 4]
```diff
@@ -69,27 +69,25 @@ export abstract class Behavior<A> extends Reactive<A, BListener>
...
-        this.last = newValue;
-        pushToChildren(t, this);
-      }
+    if (this.changedAt === t) {
+      // This prevents a second push with same timestamp, to be further pushed
+      return;
+    }
+    this.pull(t);
+    if (this.changedAt === t && this.state === State.Push) {
+      pushToChildren(t, this);
```
Comment (paldepind) [inline src/behavior.ts]: The old `pushB` updated both `changedAt` and `pulledAt`. Would it be better to still do that such that it doesn't become the responsibility of the `pull` methods?

File: src/behavior.ts [hunk 5]
```diff
@@ -446,11 +436,11 @@ class SwitcherBehavior<A> extends ActiveBehavior<A>
     this.b = newB;
     this.parents = cons(newB);
     newB.addListener(this.bNode, t);
-    const newState = newB.state;
-    if (newState !== this.state) {
-      this.changeStateDown(newState);
+    this.changeStateDown(newB.state);
```
Comment (paldepind) [inline src/behavior.ts]: Why is `changeStateDown` always called instead of only when `newState !== this.state` as before:

```ts
const newState = newB.state;
if (newState !== this.state) {
  this.changeStateDown(newB.state);
}
```

File: src/common.ts [hunk 6]
```diff
@@ -49,7 +49,9 @@ export class PushOnlyObserver<A> implements BListener, SListener<A> {
     }
   }
   pushB(t: number): void {
-    this.callback((this.source as any).last);
+    const b = <Behavior<A>>this.source;
+    b.pull(t);
+    this.callback(b.last);
```
Comment (paldepind) [inline src/common.ts]: Why this change adding a `pull`? The tests pass without this change?

File: src/common.ts [hunk 7]
```diff
@@ -145,7 +147,9 @@ export class CbObserver<A> implements BListener, SListener<A> {
...
```
Comment (paldepind) [inline src/common.ts]: Same as above, why add a `pull`? The tests pass without it.

File: src/placeholder.ts [hunk 8]
```diff
@@ -34,18 +35,14 @@ export class Placeholder<A> extends Behavior<A> {
       (<any>child).pushS(t, a);
     }
   }
-  pull(t: number): void {
+  pull(t: number) {
     if (this.source === undefined) {
       throw new SamplePlaceholderError(this);
     } else if (isBehavior(this.source)) {
+      this.source.pull(t);
```
Comment (paldepind) [inline src/placeholder.ts]: Why is the pull here no longer wrapper in `if (this.source.pulledAt !== t) {`?

File: test/behavior.ts [hunk 9]
```diff
@@ -283,6 +284,20 @@ describe("behavior", () => {
...
+        //  b1   b2
+        //   \   /
+        //    b3
+        const p = sinkBehavior(3);
+        const b1 = p.map((n) => n * n);
+        const b2 = p.map((n) => n + 4);
+        const b3 = H.lift((n, m) => n + m, b1, b2);
+        const cb = subscribeSpy(b3);
+        p.newValue(4);
+        assert.deepEqual(cb.args, [[16], [24]]);
```
Comment (paldepind) [inline test/behavior.ts]: Nice!

File: src/behavior.ts (hunk 5 above)
Comment (limemloh) [reply src/behavior.ts]: because this is already being checked inside `this.changeStateDown`

File: src/common.ts (hunk 6 above)
Comment (limemloh) [reply src/common.ts]: this is some remains of my first attempt, where the pull was not trigger during the traversal of push. I will remove this again :)

File: src/behavior.ts (hunk 5 above)
Comment (paldepind) [reply src/behavior.ts]: I see. Great change then :+1:
